test/comparison-tests/projectReferencesWatch/expectedOutput-3.6/patch1/output.txt

```diff
@@ -0,0 +1,10 @@
+    Asset      Size  Chunks             Chunk Names
+bundle.js  4.33 KiB    main  [emitted]  main
+Entrypoint main = bundle.js
+[./app.ts] 169 bytes {main} [built] [1 error]
+[./lib/index.ts] 104 bytes {main}
+
+ERROR in app.ts
```

Comment:
Is this bug in the test runner ? When I run locally same scenario and edit files, the error goes away and bundle has correct contents..

![image](https://user-images.githubusercontent.com/8052792/65080192-982c5300-d955-11e9-9c5a-3228f1f6f18c.png)

Reply:
@johnnyreilly for help

Reply:
It's possible that it is a bug, yes. The nature of the comparison tests is that they're a little flaky / imperfect. I've always meant to looking into replacing them with something slightly more solid but just never had the time.

Sorry if that's impacting you.

I inherited the comparison test infrastructure when I picked up ts-loader. The flakiness aspect is one of the reasons I quickly went on to write the execution tests which are more dependable / reliable. The downside of them is you can't test error scenarios very well.

If you've ideas / suggestions then please feel free to implement. That is always welcome

Reply:
Other note: perhaps this test might benefit from being a "watch" comparison test: https://github.com/TypeStrong/ts-loader/blob/master/test/comparison-tests/README.md#watch-specific-tests

Reply:
@johnnyreilly that's what this test is? did I miss something. I did create patches to be applied and that's where it went wrong

Reply:
Sorry - I didn't notice that. Yeah you may be bumping on a problem with the comparison test infrastructure.

Reply:
`yarn run comparison-tests -- --single-test projectReferencesWatch`

Reply:
I deleted the `expectedOutput-3.6` folder in case there was any stale files inside and regenerated the output with this command:

`yarn run comparison-tests -- --save-output --single-test projectReferencesWatch`

Reply:
You say running locally it works.  Let me try:

```
# pwd is C:\source\ts-loader
cd .\test\comparison-tests\projectReferencesWatch\
# pwd is C:\source\ts-loader\test\comparison-tests\projectReferencesWatch
node ..\..\..\node_modules\webpack\bin\webpack -w
```

Reply:
I've been thinking about this.  I think the problem may lie in `patch0` somehow.  Follow my train of thought:

The comparison test does not introduce errors along the way.  It works like this: 
1. set up legitimate mini typescript project
2. in `patch0`, updated the project reference `lib/index.ts` to include a number `four`.
3. update `app.ts` to start consuming the number `four`.

The error we're presented with in `patch1` is this: 

```
ERROR in app.ts
./app.ts
[90m[tsl] [39m[1m[31mERROR[39m[22m[1m[31m in [39m[22m[1m[36mapp.ts(3,46)[39m[22m
[1m[31m      TS2339: Property 'four' does not exist on type '{ one: number; two: number; three: number; }'.[39m[22m
```

Which suggests `patch0` didn't successfully copy over `lib/index.ts`

Reply:
This seems issue with `copySync` keeping old timestamps.. The timestamp of ts file is older than generated js file and hence the reference is not built and error shown
```
c:\github\ts-loader\test\comparison-tests\projectReferencesWatch>dir C:\github\ts-loader\.test\projectReferencesWatch\lib\index.ts C:\github\ts-loader\.test\projectReferencesWatch\lib\index.js
 Volume in drive C is OSDisk
 Volume Serial Number is F41A-DBC9

 Directory of C:\github\ts-loader\.test\projectReferencesWatch\lib

09/17/2019  12:31 PM                84 index.ts

 Directory of C:\github\ts-loader\.test\projectReferencesWatch\lib

09/18/2019  12:02 PM               137 index.js
               2 File(s)            221 bytes
               0 Dir(s)  806,273,327,104 bytes free

c:\github\ts-loader\test\comparison-tests\projectReferencesWatch>type C:\github\ts-loader\.test\projectReferencesWatch\lib\index.ts C:\github\ts-loader\.test\projectReferencesWatch\lib\index.js

C:\github\ts-loader\.test\projectReferencesWatch\lib\index.ts


export const lib = {
  one: 1,
  two: 2,
  three: 3,
  four: 4 // Add new number
};

C:\github\ts-loader\.test\projectReferencesWatch\lib\index.js


"use strict";
exports.__esModule = true;
exports.lib = {
    one: 1,
    two: 2,
    three: 3
};
//# sourceMappingURL=index.js.map
c:\github\ts-loader\test\comparison-tests\projectReferencesWatch>                                                     
```

Reply:
I tweaked the test runner to slow it down (applies patches every 10 seconds) like this:

```
function copyPatchOrEndTest(testStagingPath, watcher, testState, done) {
    // check for new files to copy in
    const patchPath = path.join(testStagingPath, 'patch' + testState.iteration);
    if (fs.existsSync(patchPath)) {
        testState.iteration++;

        console.info("patchPath: " + patchPath)

        // can get inconsistent results if copying right away
        setTimeout(function () {
            fs.copySync(patchPath, testStagingPath, { overwrite: true });
        }, 10000);
    }
    else {
        // ...
}
```

This allows me to watch the comparison test as it works.  What it's doing I can't understand right now.  It's like this:

![cant-explain](https://user-images.githubusercontent.com/1010525/65178124-fdfd0580-da4f-11e9-9ed4-59a959664f50.gif)

Reply:
Actually - it turns out me slowing it down that much messes up the test.  If I run it at normal speed it seems to be doing the right thing.  But it just doesn't seem to be picking up the changes.....

Reply:
I think maybe the comparison test runner is fine.  I think this might be a bug. It looks as though the `lib/index.ts` file is being patched, the `lib/index.d.ts` file is not being regenerated.  Take a look:

![actually-maybe-it-is-a-bug](https://user-images.githubusercontent.com/1010525/65179030-e888db00-da51-11e9-822f-76e090959233.gif)

Reply:
Yeah - I was able to repeat this locally.  Whilst webpack *is* emitting `lib/index.d.ts`, `lib/index.d.ts` has *not* be recompiled.  It never gets updated from the initial compilation:

![the-d-ts-files-are-not-regenerated](https://user-images.githubusercontent.com/1010525/65179828-7addae80-da53-11e9-9cc7-b47dc8a03313.gif)

Reply:
Sorry I missed your comment about `copySync` timestamps before.

When I reproduced locally I copied the exact file over - would that be the same behaviour?

Reply:
Fixed it :)